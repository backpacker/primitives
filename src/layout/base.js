import React from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

import { LayoutStylePropTypes } from '../prop-types';
import { makeStyle } from '../utils/style';

const BaseLayout = (props) => {
  const { children, center, onLayout, testID, ...rest } = props;

  const baseStyle = makeStyle([
    center && {
      justifyContent: 'center',
      alignItems: 'center'
    },
    { ...rest }
  ]);

  return (
    <View style={baseStyle} onLayout={onLayout} testID={testID}>
      {children}
    </View>
  );
};

BaseLayout.propTypes = {
  ...LayoutStylePropTypes,
  center: PropTypes.bool,
  onLayout: PropTypes.func
};

BaseLayout.defaultProps = {
  center: false,
  onLayout: () => {}
};

export default BaseLayout;
