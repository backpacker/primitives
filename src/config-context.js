import React, { useContext, useMemo, useState } from 'react';
import PropTypes from 'prop-types';

import baseConfig from './config';

const ConfigContext = React.createContext(undefined);

function ConfigProvider({ config: customConfig, defaultTheme, children }) {
  const [activeTheme, setActiveTheme] = useState(defaultTheme);

  const theme = useMemo(() => {
    const config = { ...baseConfig, ...customConfig };

    const baseTheme = baseConfig.default;

    const defaultSpacerSize =
      config[activeTheme]?.defaultSpacerSize || baseTheme.defaultSpacerSize;
    const spacerUnit = config[activeTheme]?.spacerUnit || baseTheme.spacerUnit;

    const spacing = (size = defaultSpacerSize) => size * spacerUnit;

    return {
      ...config[activeTheme],
      spacing
    };
  }, [activeTheme, customConfig]);

  const value = {
    theme,
    setTheme: setActiveTheme
  };

  return (
    <ConfigContext.Provider value={value}>{children}</ConfigContext.Provider>
  );
}

ConfigProvider.propTypes = {
  config: PropTypes.shape({}),
  defaultTheme: PropTypes.string
};

ConfigProvider.defaultProps = {
  config: {},
  defaultTheme: 'default'
};

const withTheme = (Component) => (props) =>
  (
    <ConfigContext.Consumer>
      {(value) => <Component {...props} {...value} />}
    </ConfigContext.Consumer>
  );

function useTheme() {
  const ctx = useContext(ConfigContext);
  if (!ctx) {
    throw new Error(`useTheme() must be called within ConfigContext`);
  }
  return ctx;
}

export { ConfigProvider, withTheme, useTheme };
export default ConfigContext;
