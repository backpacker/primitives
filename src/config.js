export const weights = {
  extralight: '100',
  thin: '200',
  light: '300',
  normal: '400',
  medium: '500',
  semibold: '600',
  bold: '700',
  heavy: '800',
  extraheavy: '900'
};

const defaultTheme = {
  fontSize: {
    h1: 32,
    h2: 24,
    h3: 20,
    h4: 16,
    h5: 14,
    h6: 12,
    h7: 10,
    h8: 8
  },
  fontWeight: weights.normal,
  fontFamily: undefined,
  weights,

  spacerUnit: 8,
  defaultSpacerSize: 2,
  spacer: {
    xs: 10,
    s: 14,
    m: 18,
    l: 22,
    xl: 26
  },

  colors: {
    brand: '#28D9AA',
    background: '#FFF',
    text: '#000'
  },

  isDark: false
};

const darkTheme = {
  ...defaultTheme,
  colors: {
    brand: '#28D9AA',
    background: '#000',
    text: '#FFF'
  },
  isDark: true
};

export { defaultTheme };

export default {
  activeTheme: 'default',
  default: defaultTheme,
  dark: darkTheme
};
