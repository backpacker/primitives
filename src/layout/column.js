import React from 'react';

import BaseLayout from './base';

const Column = (props) => {
  const { children, ...rest } = props;

  return (
    <BaseLayout flexDirection='column' {...rest}>
      {children}
    </BaseLayout>
  );
};

Column.propTypes = BaseLayout.propTypes;
Column.defaultProps = BaseLayout.defaultProps;

export default Column;
