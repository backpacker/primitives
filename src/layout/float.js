import React from 'react';

import BaseLayout from './base';

const Float = (props) => {
  const { children, ...rest } = props;

  return (
    <BaseLayout position='absolute' zIndex={1} {...rest}>
      {children}
    </BaseLayout>
  );
};

Float.propTypes = BaseLayout.propTypes;
Float.defaultProps = BaseLayout.defaultProps;

export default Float;
