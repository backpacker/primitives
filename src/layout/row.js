import React from 'react';

import BaseLayout from './base';

const Row = (props) => {
  const { children, ...rest } = props;

  return (
    <BaseLayout flexDirection='row' {...rest}>
      {children}
    </BaseLayout>
  );
};

Row.propTypes = BaseLayout.propTypes;
Row.defaultProps = BaseLayout.defaultProps;

export default Row;
