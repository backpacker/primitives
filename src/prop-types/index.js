/**
 * Style PropTypes
 */

import LayoutStylePropTypes from './layout-style';
import TouchableStylePropTypes from './touchable-style';
import TextStylePropTypes from './text-style';
import ModalPropTypes from './modal';

export {
  LayoutStylePropTypes,
  TouchableStylePropTypes,
  TextStylePropTypes,
  ModalPropTypes
};
