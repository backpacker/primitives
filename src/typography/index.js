import React, { useContext } from 'react';

import Text from './base';
import ConfigContext from '../config-context';
import config from '../config';

const makeHeadingComponent = ({ key }) => {
  return ({ children, ...rest }) => {
    const { theme } = useContext(ConfigContext);
    const { fontSize } = theme;

    return (
      <Text fontSize={fontSize[key]} {...rest}>
        {children}
      </Text>
    );
  };
};

const [H1, H2, H3, H4, H5, H6, H7, H8] = Object.entries(
  config.default.fontSize
).map(([key]) => makeHeadingComponent({ key }));

export { Text, H1, H2, H3, H4, H5, H6, H7, H8 };
