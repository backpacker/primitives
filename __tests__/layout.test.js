/* eslint-env jest */
import React from 'react';
import { render } from '@testing-library/react-native';

import { ConfigProvider } from '../src/config-context';
import { Column, Row, Float } from '../src/layout';
import { Text } from '../src/typography';

describe('<Column>', () => {
  it('should render two Text elements in column layout, centered', () => {
    const tree = (
      <ConfigProvider>
        <Column center testID='column'>
          <Text>Brian Kernighan</Text>
          <Text>Dennis Ritchie</Text>
        </Column>
      </ConfigProvider>
    );
    const { getByText, queryByTestId } = render(tree);
    const col = queryByTestId('column');
    expect(col).toHaveStyle({
      justifyContent: 'center',
      alignItems: 'center',
      flexDirection: 'column'
    });
    const t1 = getByText('Brian Kernighan');
    const t2 = getByText('Dennis Ritchie');
    expect(t1).toBeTruthy();
    expect(t2).toBeTruthy();
  });
});

describe('<Row>', () => {
  it('should render two Text elements in row layout', () => {
    const tree = (
      <ConfigProvider>
        <Row testID='row'>
          <Text>Brian Kernighan</Text>
          <Text>Dennis Ritchie</Text>
        </Row>
      </ConfigProvider>
    );
    const { getByText, queryByTestId } = render(tree);
    const col = queryByTestId('row');
    expect(col).toHaveStyle({
      flexDirection: 'row'
    });
    const t1 = getByText('Brian Kernighan');
    const t2 = getByText('Dennis Ritchie');
    expect(t1).toBeTruthy();
    expect(t2).toBeTruthy();
  });
});

describe('<Float>', () => {
  it('should render two Text elements in float layout', () => {
    const tree = (
      <ConfigProvider>
        <Float testID='row'>
          <Text>Brian Kernighan</Text>
          <Text>Dennis Ritchie</Text>
        </Float>
      </ConfigProvider>
    );
    const { getByText, queryByTestId } = render(tree);
    const col = queryByTestId('row');
    expect(col).toHaveStyle({
      position: 'absolute',
      zIndex: 1
    });
    const t1 = getByText('Brian Kernighan');
    const t2 = getByText('Dennis Ritchie');
    expect(t1).toBeTruthy();
    expect(t2).toBeTruthy();
  });
});
