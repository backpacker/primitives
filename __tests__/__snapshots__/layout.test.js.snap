// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Column> should render two H1 elements in column layout, centered 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Brian Kernighan
  </Text>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Dennis Ritchie
  </Text>
</View>
`;

exports[`<Float> should render two H1 elements in float layout 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "position": "absolute",
      "zIndex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Brian Kernighan
  </Text>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Dennis Ritchie
  </Text>
</View>
`;

exports[`<Row> should render two H1 elements in row layout 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Brian Kernighan
  </Text>
  <Text
    style={
      Object {
        "color": "#000",
        "fontFamily": undefined,
        "fontSize": 32,
        "textAlign": "left",
      }
    }
  >
    Dennis Ritchie
  </Text>
</View>
`;
